### ISSUE TITLE
How is DoS from too-large handshake messages prevented?

### ISSUE THREAD

**Issue by briansmith** [ctz/rustls#698]
[Edit: I moved the zero-length message concern to #717.]

I was reading HandshakeJoiner and MessageDeframer and it isn't clear how we prevent DoS, especially DoS caused by a peer (especially an unauthenticated peer) forcing us to allocate a huge amount of memory.

`ClientHello` and `ServerHello` messages should be required to be small. Most implementations I'm aware of require these to be quite small as sending a Handshake message header that indicates a huge Hello would be a very cheap way to encourage a server to keep a connection open for a long time. This should be enforced in the handshake joiner as soon as the handshake message length bytes are read, AFAICT. In particular, this must be enforced before huge amounts of memory are allocated for the fragments and/or joined message.

Huge handshake messages should be rejected. Probably each type of handshake message should have a different maximum length enforced for it. In particular, this must be enforced before huge amounts of memory are allocated for the fragments and/or joined message.

For handshake message sizes, I think `HandshakeJoiner` needs to do all the enforcement itself, to protect against being forced to allocate a ton of memory. In turn, the deframer probably needs to peek at the message being deframed and, if it is a handshake message, get some info from the `HandshakeJoiner` about whether to stop early.

WDYT?

**Comment by djc** [comment]
I think this makes sense. Maybe it would be easier to do this (well) after the deframer and the handshake joiner have been merged, as discussed in #668?

**Comment by briansmith** [comment]
Note that whatever maximum limits we start enforcing, we should probably assume Rustls will allocate 4x or so of memory in processing a message of that length because of the cloning of message contents it is doing currently. Reducing that factor is a separate issue to work on after we have some enforcement of message sizes.

It is likely we can enforce pretty loose limits (64K would be a good start) for most handshake message types, without any worries about compatibility. The exception I can think of off the top of my head would be certificates messages. For these, we should consider having a configuration setting to let the user choose the balance of the maximum that is enforced; this can be done separately on top of the initial work.

Note that Rustls enforces a limit on certificate sizes already:

```
        // 64KB of certificates is plenty, 16MB is obviously silly
        codec::read_vec_u24_limited(r, 0x10000)
```

The problem is that this limit seems to be enforced too late, AFAICT. However, The good news is that if we're already limiting Certificate messages to 64K then we should be able to limit *all* handshake messages to 64KB to start.

So, I suggest:
1. Change the deframer and handshake joiner so that we'll immediately reject any handshake message that would be more than 64K.
2. Release.
3. Fine tune things from there.

**Comment by jsha** [comment]
Perhaps relatedly, I noticed that `ChunkVecBuffer::len()` has O(n) performance in the number of chunks:

https://github.com/ctz/rustls/blob/beef0be1658cd457d4f47039abf4f4bb37d59a78/rustls/src/vecbuf.rs#L39-L58

It's not clear to me whether an attacker could exploit this to use disproportionate CPU by sending many small chunks.

**Comment by ctz** [comment]
Addressed in e0170ea

Note that:

> this must be enforced before huge amounts of memory are allocated for the fragments and/or joined message.

This wasn't the previous behaviour: send 5 bytes introducing a 16MB handshake message didn't cause 16MB to be immediately allocated; only sending the 16MB of data had that effect. So there was relatively little "mechanical advantage" to be gained by an attacker.
